common/kotlinx-coroutines-core-common/src/JobSupport.kt

```diff
@@ -838,89 +955,106 @@ internal open class JobSupport constructor(active: Boolean) : Job, SelectClause0
         val state = this.state
         return when (state) {
             is Finishing -> buildString {
-                if (state.cancelled != null) append("Cancelling")
-                if (state.completing) append("Completing")
+                when { // cancelling implies failing
+                    state.isCancelling -> append("Cancelling")
+                    state.isFailing -> append("Failing")
+                    else -> append("Active")
+                }
+                if (state.isCompleting) append("Completing")
             }
             is Incomplete -> if (state.isActive) "Active" else "New"
             is Cancelled -> "Cancelled"
-            is CompletedExceptionally -> "CompletedExceptionally"
+            is CompletedExceptionally -> "Failed"
             else -> "Completed"
         }
     }
 
-    // Cancelling or Completing
+    // Completing, Failing, Cancelling states,
+    // All updates are guarded by synchronized(this), reads are volatile
     @Suppress("UNCHECKED_CAST")
     private class Finishing(
         override val list: NodeList,
-        @JvmField val cancelled: Cancelled?, /* != null when cancelling */
-        @JvmField val completing: Boolean /* true when completing */
+        @Volatile
+        @JvmField var isCancelling: Boolean,
+        @Volatile
+        @JvmField var isCompleting: Boolean,
+        @Volatile
+        @JvmField var rootCause: Throwable? // NOTE: rootCause is kept even when SEALED
     ) : SynchronizedObject(), Incomplete {
-        override val isActive: Boolean get() = cancelled == null
+        @Volatile
```

Comment:
Dances are still needed because HotSpot cannot optimized multiple volatile reads even under synchronization, so you have to manually make sure you read it only once. On the other hand, we could have got rid of `@Volatile` (there is only one unsynchronized read and it can survive data race), but it is just generally safer and more future-proof to leave it volatile.